documentation/docs/04-advanced-routing.md

```diff
@@ -63,6 +63,12 @@ export function load(event) {
 
 > If you don't handle 404 cases, they will appear in [`handleError`](/docs/hooks#shared-hooks-handleerror)
 
+### Optional parameters
+
+A route like `[lang]/home` contains a parameter named `lang` which is required. Sometimes it's beneficial to make these parameters optional, so that in this example both `home` and `en/home` point to the same page. You can do that by wrapping the parameter in another bracket pair: `[[lang]]/home`
+
+Note that an optional route parameter cannot follow a rest parameter (`[...rest]/[[optional]]`), as it's impossible to know which way to match the route.
```

Comment:
Does `[[optional]]/[...rest]` work? I would find it weird if one worked but not the other. You'd run into the same ambiguities either way. I'd expect both `[[optional]]`s to be greedy, which would probably mean they only make sense if you are using a route matcher.

Reply:
@Rich-Harris Do you have thoughts about this? I see your solution was "disallow both", which is certainly _a_ solution, and probably the simplest one.

Reply:
It's the simplest, but it's also the most future-proof, since if someone presents a compelling use case we can consider allowing them next to each other in a future release in a non-breaking way. Possibly by (for example) only allowing `[[optional=somematcher]]`

Reply:
Makes sense to me. I'm good with disallowing both usages.

Reply:
The real problem is that the regex is greedy, so one works but not the other. I'm not sure if this is a good argument to allow optional before rest.

Reply:
```suggestion
All route matches are done greedy. That means that a route like `[[optional1]]/[[optional2]]` matched against `/foo` will _always_ result in `{ optional1: 'foo' }`. For this reason an optional route parameter cannot follow a rest parameter (`[...rest]/[[optional]]`), as the `optional` parameter will never be matched.
```

Reply:
(I've updated "greedy" to "greedily" in your suggestion.)

But I'm not really sure it's accurate to say that we _can't_ do this - it's just true that we can't do this with our current method of generating regexes for the route manifest. There's nothing stopping us from using non-greedy regexes or (in the `[...foo]/[[bar=matcher]]` case) needing to rely on a matcher to know which route should match. It may well be that we make the decision that these routes aren't worth the hassle of supporting or that they would result in a performance hit because we can't rely on regexes as much - but we should be careful not to frame these in the documentation as being inherently impossible to support.

Reply:
I changed the implementation (and the docs) to disallow ambiguity in both directions